services/replicator/metadataReconciler.go

```diff
@@ -265,6 +286,121 @@ func (r *metadataReconciler) reconcileDest(localDests []*shared.DestinationDescr
 	r.m3Client.UpdateGauge(metrics.ReplicatorReconcileScope, metrics.ReplicatorReconcileDestFoundMissing, replicatorReconcileDestFoundMissingCount)
 }
 
+func (r *metadataReconciler) reconcileCg(localCgs []*shared.ConsumerGroupDescription, remoteCgs []*shared.ConsumerGroupDescription) {
+	var replicatorReconcileCgFoundMissingCount int64
+	localCgsSet := make(map[string]*shared.ConsumerGroupDescription)
+	for _, cg := range localCgs {
+		localCgsSet[cg.GetConsumerGroupUUID()] = cg
+	}
+
+	for _, remoteCg := range remoteCgs {
+		lclLg := r.logger.WithFields(bark.Fields{
+			common.TagCnsm: common.FmtCnsm(remoteCg.GetConsumerGroupUUID()),
+			common.TagDst:  common.FmtDst(remoteCg.GetDestinationUUID()),
+		})
+		localCg, ok := localCgsSet[remoteCg.GetConsumerGroupUUID()]
+		if ok {
+			if remoteCg.GetStatus() == shared.ConsumerGroupStatus_DELETED {
+				// case #1: ConsumerGroup gets deleted in remote, but not deleted in local. Delete the ConsumerGroup locally
+				if !(localCg.GetStatus() == shared.ConsumerGroupStatus_DELETED) {
+					lclLg.Info(`Found deleted cg from remote but not deleted locally`)
+					deleteRequest := &shared.DeleteConsumerGroupRequest{
+						DestinationUUID:   common.StringPtr(remoteCg.GetDestinationUUID()),
+						ConsumerGroupName: common.StringPtr(remoteCg.GetConsumerGroupName()),
+					}
+					ctx, cancel := thrift.NewContext(localReplicatorCallTimeOut)
+					defer cancel()
+					err := r.replicator.DeleteConsumerGroup(ctx, deleteRequest)
+					if err != nil {
+						lclLg.Error(`Failed to delete ConsumerGroup in local zone for reconciliation`)
+						continue
+					}
+				} else {
+					lclLg.Info(`Found ConsumerGroup is deleted in both remote and local`)
+					continue
+				}
+				continue
+			}
+
+			// TODO case #2: ConsumerGroup exists in both remote and local, try to compare the property to see if anything gets updated
+		} else {
+			// case #3: ConsumerGroup exists in remote, but not in local. Create the ConsumerGroup locally
```

Comment:
[ Not a Blocker ] Given that this is a state machine, one way to avoid nested if else is to use a table driven approach of state handlers. For example, if you have a table like the following:

stateHandlers[localCG.Status][remoteCG.status] -> func(localCG, remoteCG)

Then, you just have to index into this table to invoke the appropriate handler for the states.